python/cog/server/redis_queue.py

```diff
@@ -160,13 +160,29 @@ def start(self) -> None:
             sys.stderr.write(f"Setup time: {setup_time:.2f}\n")
 
         sys.stderr.write(f"Waiting for message on {self.input_queue}\n")
+        consecutive_failures = 0
         while not self.should_exit:
             try:
                 message_id, message_json = self.receive_message()
-                if message_json is None:
-                    # tight loop in order to respect self.should_exit
+                consecutive_failures = 0
+            except:
+                # Give up and exit after a few retries. The cost of running
+                # `setup()` is significant, so we don't want to exit due to an
+                # intermittent error. But, we do want to crash if we can't
+                # connect to Redis, so the failure becomes evident.
+                consecutive_failures += 1
+                if consecutive_failures > 10:
+                    sys.stderr.write("Failed to receive message from Redis; giving up")
+                    break
+                else:
+                    sys.stderr.write("Failed to receive message from Redis; retrying")
                     continue
```

Comment:
A short sleep would be nice in case of an intermittent blip. Backoffs are probably not necessary – in the case where the Cog model is deployed on Kubernetes, the cluster will CrashLoopBackoff.

Reply:
I was thinking about the case where we've got a model that takes a couple of minutes to run `setup()`, and Redis goes down for 30s. We'd prefer to keep the model alive through that, rather than have to restart it and rerun `setup()`, I think. Weighed against: if there's a problem with config, which means it'll never be able to connect to Redis, we want it to crash sooner rather than later.

I wonder about doing a connection check before running `setup()`, so we can fail quickly, and then allowing rather longer before failing after `setup()` is complete.